Module1

sonasol2 opened the pull request:
1. added projects DAL and Domain 
2. added validation in API layer. Removed business logic from controller
3. added constant. Removed magic value
4. removed comments
5. added parser. 
6. In domain create advertising service where moved business logic
7. added based repository and child repository

sonasol2 committed:
update structure and clearing code

update:
- rename solution from AdRegion to AdvertisingRegionService
- rename projet AdRegion to AdvertisingRegionService.API
- rename project AdRegion.Test to AdvertisingRegionService.Test
- clean code from comments

add:
- add new projects AdvertisingRegionService.DAL and AdvertisingRegionService.Domain

sonasol2 committed:
validate and constants fix

add:
-Constants for message and other value
-validators
-validator service
-integrate fluent validation

update:
-replaced magic value with constants
-remove validation from service
-replaced validation in controller

sonasol2 committed:
configuration segregation

add:
-configuration application file
-configuration service file

update:
-moved configurations from program.cs to new added config files

sonasol2 committed:
update

add:
- AdRegionFileParser and IAdRegionFileParser

update:
- moved all parse process from service to parser
- add more constants
- fix validate param
- register parse service

sonasol2 committed:
integrate repository

add:
-IRepository
-ICacheRepository and CacheRepository
-IEntity
-AdvertisingPlatform entity
-CacheContext - improvised database

Update:
-parser is work with an entity AdvertisingPlatform
-add some catch exception to AdRegionService
-configuration(repository registration)

Other fix:
-fix project reference
-fix namespace
-clear some usings

sonasol2 committed:
fix

Fix
-project reference
-namespace
-clear some using
-Constants usage

Vadya0994 commented on AdvertisingRegionService.API/Controllers/AdPlatformsController.cs:
```diff
@@ -0,0 +1,53 @@
+    [Route("upload")]
+    public async Task<IActionResult> UploadFile(IFormFile? request)
+    {   
+        var validatorResult = await _validatorService.ValidateAsync(request);
```
Вот это все можно в мидлваре встроить одной строчкой кода, вместо постоянного вызова в каждом методе контроллера.

Vadya0994 replied:
https://www.bytehide.com/blog/fluent-validation-asp-net-core - вот тут есть примеры

sonasol2 replied:
Добавил как TODO чтобы обговорить этот вопрос на созвоне

Vadya0994 commented on AdvertisingRegionService.DAL/Repositories/CacheRepository.cs:
```diff
@@ -0,0 +1,110 @@
+        {
+            _context.AdvertisingPlatform.Remove(entity);
+        }
+        catch (Exception e)
```
отлов общих эксепшенов можно встроить на уровень мидлваре, т.е написать небольшую мидлвару для обработки эксепшенов. Так количество кода заметно сократится)

sonasol2 replied:
Сделал. Взял мидлвар который скинул @chaika-ds.

Vadya0994 commented on AdvertisingRegionService.Domain/Constants.cs:
```diff
@@ -0,0 +1,37 @@
+    public const int MaxSearchLength = 100;
+    public const int MinSearchLength = 2;
+    
+    //message
```
слушай, ты даже комментарием подчеркиваешь то, что напрашивается в отдельный файл)  Можешь его обозвать LocalizationConstants

sonasol2 replied:
вынес

Vadya0994 commented on AdvertisingRegionService.Domain/Constants.cs:
```diff
@@ -0,0 +1,37 @@
+    
+    public const string NullModelMessage = "Model is null";
+    
+    //parse
```
и вот эти тоже)

sonasol2 replied:
вынес

Vadya0994 commented on AdvertisingRegionService.Domain/Interfaces/IAdRegionFileParser.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace AdvertisingRegionService.Domain.Interfaces;
+
+public interface IAdRegionFileParser
```
AdRegion - я бы не делал сокращения названий в названии классов. Т.е. вместо Ad писал бы прям Advertising. И так везде по проекту.

sonasol2 replied:
поправил во всем проекте

Vadya0994 commented on AdvertisingRegionService.Domain/Interfaces/IAdRegionService.cs:
```diff
@@ -0,0 +1,10 @@
+using AdvertisingRegionService.Domain.Models;
+using Microsoft.AspNetCore.Http;
```
как-нибудь можем без этого в домене обойтись? Понимаю что тянешь IformFile оттуда. Но смотри вариантов может быть много, например поток байт достать оттуда и оперировать им. Или как вариант в Api методе просто сохранить файл на сервер и передавать название файла, и уже пользуясь стандартным функционалом открывать файл и читать его.

sonasol2 replied:
Моя мысль была в том чтобы вынести работу с файлом из контроллера.
Решил передавать Stream, и не нужные юзинги убрал:)

Vadya0994 commented on AdvertisingRegionService.Domain/Parsers/AdRegionFileParser.cs:
```diff
@@ -0,0 +1,72 @@
+                var currentRegion = region;
+                var platformsForLocation = new List<string>();
+
+                while (!string.IsNullOrEmpty(currentRegion))
```
можно в отдельный метод вынести говорящий, использование цикла в цикле усложняет понимание того что происходит

sonasol2 replied:
Сделал! За одно понял что можно сделать более грамотнее

Vadya0994 commented on AdvertisingRegionService.Domain/Services/AdRegionService.cs:
```diff
@@ -0,0 +1,46 @@
+        }
+        catch (Exception e)
+        {
+            WorkResult<HashSet<string>>.Fail(e.Message);
```
наверное return нужен?)

sonasol2 replied:
Точно, добавил:)

Vadya0994 commented on AdvertisingRegionService.Domain/Services/AdRegionService.cs:
```diff
@@ -0,0 +1,46 @@
+        {
+            return WorkResult<HashSet<string>>.Success(_repository.GetAdvertisingPlatformByLocation(searchRequest));
+        }
+        catch (Exception e)
```
Возможно тоже просится на уровень мидлвары? Там ты сможешь также в result завернуть.

sonasol2 replied:
Я сделал глобальный отлов исключений, и задумался, а есть ли вообще смысл в таком подходе, с моделью результата?

Vadya0994 commented on AdvertisingRegionService.DAL/Interfaces/ICacheRepository.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace AdvertisingRegionService.DAL.Interfaces;
+
+public interface ICacheRepository : IRepository<AdvertisingPlatform>
```
Интерфейсы и публичные модели без реализаций интерфейсов хорошо бы документировать через /// <summary> . И свойства и методы внутри них тоже. Это общепринятый формат, который поддерживают различные АПИ, так будет формироваться и выводиться в других местах документация в приятном виде. Да и в целом другие разработчики будут глядеть описания методов и свойств и возможно обращать внимание на какие-то нюансы, которые ты пропишешь.

sonasol2 replied:
Сделал в этом интерфейсе. В дальнейшем буду использовать!

chaika-ds commented on AdvertisingRegionService.API/AdvertisingRegionService.API.csproj:
```diff
@@ -0,0 +1,35 @@
+    
+    <ItemGroup>
+        <ProjectReference Include="..\AdvertisingRegionService.Domain\AdvertisingRegionService.Domain.csproj"/>
+        <ProjectReference Include="..\AdvertisingRegionService.DAL\AdvertisingRegionService.DAL.csproj"/>
```
Апи не должно ничего знать о базе данных

sonasol2 replied:
Да, он ничего не знает, просто ссылку забыл удалить:)

chaika-ds commented on AdvertisingRegionService.API/Configurations/ConfigurationApplication.cs:
```diff
@@ -0,0 +1,18 @@
+
+public static class ConfigurationApplication
+{
+    public static WebApplication ConfigureApplication(this WebApplication app)
```
Сделай метод void

sonasol2 replied:
сделал

chaika-ds commented on AdvertisingRegionService.API/Configurations/ConfigurationApplication.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public static WebApplication ConfigureApplication(this WebApplication app)
+    {
+        if (app.Environment.IsDevelopment())
```
Почему сваггер используется только для Development?

sonasol2 replied:
Я думал что он только для него и используется:))

chaika-ds commented on AdvertisingRegionService.API/Configurations/ConfigurationService.cs:
```diff
@@ -0,0 +1,54 @@
+
+public static class ConfigurationService
+{
+    public static IServiceCollection ConfigureServices(this IServiceCollection services)
```
Здесь тоже все методы лучше сделать void, здесь же только наполнется di сервисов

sonasol2 replied:
сделал

chaika-ds commented on AdvertisingRegionService.Domain/Constants.cs:
```diff
@@ -0,0 +1,37 @@
+namespace AdvertisingRegionService.Domain;
+
+public static class Constants
```
Сделай папку с константами и туда по классам разнеси константы для разных случаев использования

sonasol2 replied:
сделал

chaika-ds commented on AdvertisingRegionService.API/Dockerfile:
По докер файлу. Его нужно будет подкорректировать и закинуть на гитлаб вместе с проектом, чтоб он собрался. Это может в следующем модуле попробуем. У меня есть пример докер файла для гитлаба, посмотрим на планировании 2 модуля

sonasol2 replied:
Он у меня автоматом сгенерировался, забыл отключить, но было бы здорово с ним поработать!

chaika-ds commented on AdvertisingRegionService.API/Validators/ValidatorService.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        var validator = _serviceProvider.GetService<IValidator<T>>();
+        
+        if (validator is null)
```
Попробуй здесь применить тренарный оператор ? :
Также вместо typeof(T).Name можно насыпать сахаром nameof(T)

sonasol2 replied:
сахар насыпал! тернарный  не получается, нужно чтобы две ветки были одного типа ну и плюс там throw, а с ним вообще не работает. Может быть такой вариант сделать? Вроде лаконичнее чес было. 
```
var validator = _serviceProvider.GetService<IValidator<T>>() 
                        ?? throw new InvalidOperationException($"No validator found for type {typeof(T).Name}");
    
        return await validator.ValidateAsync(model);
```

chaika-ds commented on AdvertisingRegionService.DAL/CacheContext.cs:
```diff
@@ -0,0 +1,9 @@
+
+public class CacheContext
+{
+    public ICollection<AdvertisingPlatform> AdvertisingPlatform { get; set; } = new List<AdvertisingPlatform>();
```
Предлагаю постараться полнее симитировать бд и сделать AdvertisingPlatform статическим с синхронизацией для органичения доступа других потоков. Но это наверное на следующий модуль задачка.

sonasol2 replied:
Добавил как TODO

chaika-ds commented on AdvertisingRegionService.DAL/Configurations/ConfigurationDALService.cs:
```diff
@@ -0,0 +1,17 @@
+
+public static class ConfigurationDALService
+{
+    public static IServiceCollection AddRepositories(this IServiceCollection services)
```
Конфигурирование репозитория вынести в ко всем конфигурациям в апи проект. Разбрасывание конфигурации по проектам будет путать

sonasol2 replied:
Да я так и хотел, но, тогда как сделать так чтобы апи не знал о DAL слое? Приходит ссылку прокидывать чтобы конфиги добавить

chaika-ds commented on AdvertisingRegionService.DAL/Interfaces/IEntity.cs:
```diff
@@ -0,0 +1,6 @@
+namespace AdvertisingRegionService.Domain.Interfaces;
```
исправь namespace и проверь по всем классам правильно ли проставились namespace  при разнесении классов

sonasol2 replied:
поправил

chaika-ds commented on AdvertisingRegionService.DAL/Models/AdvertisingPlatform.cs:
```diff
@@ -0,0 +1,10 @@
+using AdvertisingRegionService.Domain.Interfaces;
+
+namespace AdvertisingRegionService.Domain.Models;
```
namespace

sonasol2 replied:
Поправил

chaika-ds commented on AdvertisingRegionService.DAL/Models/AdvertisingPlatform.cs:
```diff
@@ -0,0 +1,10 @@
+public class AdvertisingPlatform : IEntity
+{
+    public string Region { get; set; }
+    public HashSet<string> Platforms { get; set; } // подумать над тем какую коллекцию использовать
```
Такие места можно обозначать чере //TODO: это будет подсвечивать запланированные места под рефакторинг

sonasol2 replied:
Сделал, спасибо!

chaika-ds commented on AdvertisingRegionService.DAL/Repositories/CacheRepository.cs:
```diff
@@ -0,0 +1,110 @@
+
+    public AdvertisingPlatform GetById(Guid id)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(nameof(id));
```
Проверка прикольная, но это можно передать на валидацию входящих данных

sonasol2 replied:
Я думал, что пока не использую этот метод не заморачиваться с валидацие, а как только начну использовать то сразу сделаю:)

chaika-ds commented on AdvertisingRegionService.DAL/Repositories/CacheRepository.cs:
```diff
@@ -0,0 +1,110 @@
+                throw new ArgumentNullException(nameof(entities));
+            }
+
+            if (_context.AdvertisingPlatform is List<AdvertisingPlatform> list)
```
Тут не совсем понятно, в какиз случаях AdvertisingPlatform не будет листом, плюc там все равно под капотом метод Add живет

sonasol2 replied:
Я подумал так сделать чтобы избежать цикла, если разницы нету то конечно уберу

chaika-ds commented on AdvertisingRegionService.DAL/Repositories/CacheRepository.cs:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    public void AddRange(ICollection<AdvertisingPlatform> entities)
```
Нужно больше конкретизировать названия параметров и переменных.
Например не entities а в данном случае аdvertisingPlatforms. В таком же ключе переименуй list и _context лучше конкретизировать в _fileCacheContext

sonasol2 replied:
сделал

chaika-ds commented on AdvertisingRegionService.DAL/Repositories/CacheRepository.cs:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+    
+    public void Delete(AdvertisingPlatform entity)
```
Если ты внутри используешь Remove, то почему бы метод не назвать RemoveEntity

sonasol2 replied:
Логично, но почему-то привычнее делать remove:)) поправил

chaika-ds commented on AdvertisingRegionService.DAL/Repositories/CacheRepository.cs:
```diff
@@ -0,0 +1,110 @@
+
+    public HashSet<string> GetAdvertisingPlatformByLocation(string location)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(location);
```
Тут тоже проверку желательно в валидацию вынести

sonasol2 replied:
Эта проверка у меня уже есть в валидации, зачем дублировал, не знаю:))

chaika-ds commented on AdvertisingRegionService.Domain/Models/WorkResult.cs:
```diff
@@ -1,4 +1,4 @@
-namespace AdRegion.Models;
+namespace AdvertisingRegionService.Domain.Models;
 
 public class WorkResult<T>
```
Тут просится название ExecutionResult

sonasol2 replied:
поправил

chaika-ds commented on AdvertisingRegionService.Domain/Parsers/AdRegionFileParser.cs:
```diff
@@ -0,0 +1,72 @@
+                    var platform = parts[0].Trim();
+                    var region = parts[1].Split(Constants.RegionSplitter).Select(l => l.Trim().TrimEnd()).ToList();
+                    
+                    foreach (var location in region) 
```
Если цикл в цикле, то вложенный рекомендуют выносить в отдельный метод для упрощения чтения кода и возможного переиспользования

sonasol2 replied:
Вынес в отдельный метод

chaika-ds commented on AdvertisingRegionService.Domain/Parsers/AdRegionFileParser.cs:
```diff
@@ -0,0 +1,72 @@
+                platforms.Add(new AdvertisingPlatform()
+                {
+                    Region = region,
+                    Platforms = platformsForLocation.Distinct().ToList().ToHashSet()
```
Зачем здесь два преобразования подряд ToList и ToHashSet?

sonasol2 replied:
видать увлекся:)) поправил

chaika-ds commented on AdvertisingRegionService.Domain/Services/AdRegionService.cs:
```diff
@@ -0,0 +1,46 @@
+    public AdRegionService(IAdRegionFileParser parser, ICacheRepository repository)
+    {
+        _parser = parser;
+        _repository = repository;
```
Конкретизируй что за парсер, что за репозиторий

sonasol2 replied:
Конкретизировал

chaika-ds commented on AdvertisingRegionService.Domain/Services/AdRegionService.cs:
```diff
@@ -0,0 +1,46 @@
+{
+    private readonly IAdRegionFileParser _parser;
+    private static Dictionary<string, List<string>> _cachedResults = new();
+    private  readonly ICacheRepository _repository;
```
Для чтения лучше отделять свойства и параметры от конструктора строкой

sonasol2 replied:
Хорошо, сделал

chaika-ds commented on AdvertisingRegionService.Domain/Services/AdRegionService.cs:
```diff
@@ -0,0 +1,46 @@
+        using var reader = new StreamReader(file.OpenReadStream());
+        var content = await reader.ReadToEndAsync();
+
+        var platforms = _parser.Parse(content);
```
platforms  и dictionary можно поместить в один метод в парсере

sonasol2 replied:
Объединил

chaika-ds commented on AdvertisingRegionService.Domain/Services/AdRegionService.cs:
```diff
@@ -0,0 +1,46 @@
+        var dictionary = _parser.BuildRegionHierarchy(platforms);
+        
+        _repository.ClearCache();
+        _repository.AddRange(dictionary);
```
Работу с репозиторием тоже можно объединить в один метод

sonasol2 replied:
сделал

sonasol2 committed:
update

add:
-AppMiddlewareException
-ExceptionExtension

update:
-configuration methods void now
-configuration in API layer only
-separate constants
-optimisation parser
-abbreviations in methods now full name

remove:
-DAL reference from API layer

fix:
-usings
-namespaces

sonasol2 committed:
fix

remove using
